contracts/prover/PolymerProver.sol

```diff
@@ -0,0 +1,242 @@
+     * @param _portal Address of the Portal contract
+     */
+    constructor(
+        address _owner,
+        address _portal) BaseProver(_portal) Ownable(_owner) {
+    }
+
+    /**
+     * @notice Initializes the contract with CrossL2ProverV2 and whitelist settings
+     * @param _crossL2ProverV2 Address of the CrossL2ProverV2 contract
+     * @param _chainIds Array of chain IDs for whitelisted emitters
+     * @param _whitelistedEmitters Array of whitelisted emitter addresses
+     */
+    function initialize(
+        address _crossL2ProverV2,
+        uint64[] calldata _chainIds,
+        bytes32[] calldata _whitelistedEmitters) external onlyOwner {
+        if (_crossL2ProverV2 == address(0)) revert ZeroAddress();
+        if (_chainIds.length != _whitelistedEmitters.length) revert SizeMismatch();
+
+        CROSS_L2_PROVER_V2 = ICrossL2ProverV2(_crossL2ProverV2);
+
+        for (uint256 i = 0; i < _chainIds.length; i++) {
+            WHITELISTED_EMITTERS[_chainIds[i]] = _whitelistedEmitters[i];
+        }
+
+        renounceOwnership();
+    }
+
+    // ------------- STANDARD PROOF VALIDATION -------------
+
+    /**
+     * @notice Validates a single proof
+     * @param proof The proof data for CROSS_L2_PROVER_V2 to validate
+     */
+    function validate(bytes calldata proof) external {
+        (bytes32 intentHash, address claimant, uint64 destinationChainId) = _validateProof(proof);
+        processIntent(intentHash, claimant, destinationChainId);
+    }
+
+
+    /**
+     * @notice Validates multiple proofs in a batch
+     * @param proofs Array of proof data to validate
+     */
+    function validateBatch(bytes[] calldata proofs) external {
+        for (uint256 i = 0; i < proofs.length; i++) {
+            (bytes32 intentHash, address claimant, uint64 destinationChainId) = _validateProof(proofs[i]);
+            processIntent(intentHash, claimant, destinationChainId);
+        }
+    }
+
+    // ------------- INTERNAL FUNCTIONS - PROOF VALIDATION -------------
+
+    /**
+     * @notice Core proof validation logic
+     * @param proof The proof data to validate
+     * @return intentHash Hash of the proven intent
+     * @return claimant Address that fulfilled the intent
+     * @return chainId Chain ID where the event was emitted
+     */
+    function _validateProof(
+        bytes calldata proof
+    ) internal view returns (bytes32 intentHash, address claimant, uint64 chainId) {
+        (
+            uint32 destinationChainId,
+            address emittingContract,
+            bytes memory topics,
+            /* bytes memory data */
+        ) = CROSS_L2_PROVER_V2.validateEvent(proof);
+
+        if (!isWhitelisted(uint64(destinationChainId), bytes32(uint256(uint160(emittingContract))))) {
+            revert InvalidEmittingContract(emittingContract);
+        }
+        checkTopicLength(topics, 128); // 4 topics: signature + hash + claimant + sourceChainId
```

Comment:
I'd just inline this and remove the function. It's a one-liner anyway. Also, create a immutable constant for `128` please.